### PR TITLE
Add spin_and_wait_for_matched to PublicationManager and update test c…

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Address #790

### CODE REVIEW

**Comment by MichaelOrlov** [comment]
@Barry-Xu-2018 I am kind of confused about your implementation and this is not what I expected to see.
Why do you need to spin node for PublicationManager?
It does make sense to spin node only for subscription.

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp [hunk 1]
```diff
@@ -76,26 +79,78 @@ class PublicationManager
...
+    } while ((clock::now() - start) < timeout);
+
+    return false;
+  }
+
 private:
   std::shared_ptr<rclcpp::Node> pub_node_;
   using PublicationF = std::function<void (bool)>;
-  std::vector<PublicationF> publishers_;
+  std::vector<PublicationF> publishers_fcn_;
```
Comment (emersonknapp) [inline rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:152]: Is there any reason why we should maintain two separate vectors? I would prefer 

```
std::vector<std::pair<rclcpp::PublisherBase::SharedPtr, PublicationF>> publishers_;
```

This way it's impossible for the vectors to get out of sync.

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp [hunk 2]
```diff
@@ -76,26 +79,78 @@ class PublicationManager
 
   void run_publishers(bool verbose = false) const
   {
-    for (const auto & pub_fcn : publishers_) {
+    for (const auto & pub_fcn : publishers_fcn_) {
       pub_fcn(verbose);
     }
+    rclcpp::executors::SingleThreadedExecutor exec;
+    exec.spin_node_some(pub_node_);
```
Comment (emersonknapp) [inline rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:86]: Is spinning necessary for publishing?

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp [hunk 3]
```diff
@@ -76,26 +79,78 @@ class PublicationManager
...
+    std::this_thread::sleep_for(100ms);
```
Comment (emersonknapp) [inline rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:87]: Is 100ms meaningful? Where does this number come from? Why do we need to sleep? Does this make the publishing "bursty"?

**Comment by Barry-Xu-2018** [comment]
@MichaelOrlov 

> I am kind of confused about your implementation and this is not what I expected to see.
Why do you need to spin node for PublicationManager?
It does make sense to spin node only for subscription.

According to your suggestion in https://github.com/ros2/rosbag2/issues/790#issuecomment-868945959
```
For the Problem 1 related to the discovery it would be more appropriate to add wait_for_matched(..) API for the PublicationManager. Similar as made spin_and_wait_for_matched in SubscriptionManager.
```
My understanding is to add similar API like spin_and_wait_for_matched in SubscriptionManager to PublicationManager.  
Spin node is to confirm the number of connected subscription (from rosbag) for specified publisher.

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp (hunk 2 above)
Comment (Barry-Xu-2018) [reply rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:86]: Not necessary.

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp (hunk 3 above)
Comment (Barry-Xu-2018) [reply rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:87]: Not necessary. I forgot to remove them.

**Comment by MichaelOrlov** [comment]
@Barry-Xu-2018 It seems you misinterpreted my suggestion. I intentionally mentioned `wait_for_matched(..)` API without word spin. Also mentioned **`similar`** as `spin_and_wait_for_matched(..)`. I meant that `similar` doesn't means "the same as".

Anyway, thank you for your PR, I will take a look more closely on it since you have addressed some review comments.

**Comment by Barry-Xu-2018** [comment]
@MichaelOrlov 

> It seems you misinterpreted my suggestion. I intentionally mentioned wait_for_matched(..) API without word spin. Also mentioned similar as spin_and_wait_for_matched(..). I meant that similar doesn't means "the same as".

Sorry for my misunderstanding.   
Spin is  unnecessary.

> Anyway, thank you for your PR, I will take a look more closely on it since you have addressed some review comments.

Please wait a moment for the review.   
I will update codes to remove spin.

**Comment by Barry-Xu-2018** [comment]
@MichaelOrlov @emersonknapp 

Addressed your comments.   
Please review again.

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp [hunk 4]
```diff
@@ -71,31 +72,90 @@ class PublicationManager
...
+    auto sleep_time = std::chrono::milliseconds(100);
+
+    if (timeout < std::chrono::seconds(1)) {
+      sleep_time = timeout;
+    }
+
+    using clock = std::chrono::steady_clock;
+    auto start = clock::now();
+
+    rclcpp::executors::SingleThreadedExecutor exec;
```
Comment (MichaelOrlov) [inline rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:141]: It seems forgot to remove 
```
rclcpp::executors::SingleThreadedExecutor exec;
```

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp [hunk 5]
```diff
@@ -71,31 +72,90 @@ class PublicationManager
...
+      return false;
+    }
+
+    if (publisher->get_subscription_count() +
+      publisher->get_intra_process_subscription_count() >= n_subscribers_to_match)
+    {
+      return true;
+    }
+
+    auto sleep_time = std::chrono::milliseconds(100);
```
Comment (MichaelOrlov) [inline rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:132]: I would recommend to reduce default sleep timeout. I would chose some number from 1 to 10 mSec.

**Review by MichaelOrlov — COMMENTED** [review]
@Barry-Xu-2018 LGTM among a few nitpicks.

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp (hunk 4 above)
Comment (Barry-Xu-2018) [reply rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:141]: Sorry. My mistake.

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp (hunk 5 above)
Comment (Barry-Xu-2018) [reply rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp]: Okay. I change to use 10ms as default.

**Review by MichaelOrlov — APPROVED** [review]
@Barry-Xu-2018 All looks good now.
Thank you for your contribution.

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp [hunk 6]
```diff
@@ -71,31 +72,89 @@ class PublicationManager
...
   }
 
+  /// \brief Wait until publisher with specified topic will be connected to the subscribers or
+  ///    timeout occur.
+  /// \tparam Timeout Data type for timeout duration from std::chrono:: namespace
+  /// \param topic_name topic name to find corresponding publisher
+  /// \param timeout Maximum time duration during which discovery should happen.
+  /// \param n_subscribers_to_match Number of subscribers publisher should have for match.
+  /// \return true if not find publisher by topic name or publisher has specified number of
+  ///    subscribers, otherwise false.
```
Comment (MichaelOrlov) [inline rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:105]: Ups. I overlooked mistake in description for return value. It should be other way around. i.e. return false if fail and true if success. 
@Barry-Xu-2018 Could you please fix it before merge?

**Review by MichaelOrlov — CHANGES_REQUESTED** [review]

**Comment by MichaelOrlov** [comment]
Gist: https://gist.githubusercontent.com/MichaelOrlov/8f92e1829d0071ceb0844bff7ad53c52/raw/e7160639b7b34096e84da97c39d90e499def5c70/ros2.repos

BUILD args: --packages-up-to rosbag2_test_common  rosbag2_tests
TEST args: --packages-select rosbag2_test_common  rosbag2_tests
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/8641/
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14826/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9576/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12492/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14981/)

File: rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp (hunk 6 above)
Comment (Barry-Xu-2018) [reply rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp:105]: Thank you.  
The description about return value is incorrect.

**Review by MichaelOrlov — APPROVED** [review]

**Pull request merged**